larq_compute_engine/mlir/transforms/prepare_patterns_common.td

```diff
@@ -4,9 +4,18 @@ include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops.td"
 include "larq_compute_engine/mlir/ir/lce_ops.td"
 include "larq_compute_engine/mlir/transforms/op_removal_patterns.td"
 
+class ConstantValue<string val> : AttrConstraint<CPred<"IsConstantValue($_self, " # val # ")">>;
 
-// This relies on implementation details of larq.math.sign. We should make
-// this more general in the future
+// Base quantiser pattern that matches the `tf.where` implementation of `ste_sign`.
+def : Pat<(TF_SelectV2Op:$select_op
```

Comment:
This matches `tf.where(x >= 0, 1, -1)` but it does not match `tf.where(x < 0, -1, 1)` which is the same.
I think the converter should accept both forms: it can insert a `not` operation to go into a canonical form.

As far as which of the two would be the better canonical form, see also my comment in the Quantize kernel code: I think the `tf.where(x < 0, -1, 1)` makes more sense from the kernel point of view, because then a `true` is bitpacked to `1` and a `false` is bitpacked to `0`. I don't feel strongly about this, though, but I think both forms should be recognized by the converter.

Reply:
Ah that's a fair point, I hadn't realised that despite having the 'sign canonicalisers' in the optimise pass, there is still a possible issue here, since the `LceQuantise` won't even be inserted. I'll have a think about a solution.

Reply:
Should be covered by c8f25d2.